mass_mailing_custom_unsubscribe/models/mail_mass_mailing.py

```diff
@@ -31,3 +33,41 @@ def update_opt_out(self, email, list_ids, value):
                 "action": action,
             })
         return super().update_opt_out(email, list_ids, value)
+
+    def update_opt_out_other(self, email, res_ids, value):
+        model = self.env[self.mailing_model_real].with_context(
+            active_test=False)
+        action = "unsubscription" if value else "subscription"
+        if 'opt_out' in model._fields:
+            email_fname = 'email_from'
+            if 'email' in model._fields:
+                email_fname = 'email'
+            records = model.search([
+                ('id', 'in', res_ids), (email_fname, 'ilike', email)])
+            records.write({'opt_out': value})
+            for res_id in res_ids:
+                self.env["mail.unsubscription"].create({
+                    "email": email,
+                    "mass_mailing_id": self.id,
+                    "unsubscriber_id": "%s,%d" % (
+                        self.mailing_model_real, res_id),
+                    "action": action,
+                })
+
+    def _get_opt_out_list(self):
+        """Handle models with opt_out field for excluding them."""
+        self.ensure_one()
+        model = self.env[self.mailing_model_real].with_context(
+            active_test=False)
+        if (self.mailing_model_real != "mail.mass_mailing.contact" and
+                'opt_out' in model._fields):
+            email_fname = 'email_from'
+            if 'email' in model._fields:
+                email_fname = 'email'
+            domain = safe_eval(self.mailing_domain)
+            domain += [('opt_out', '=', True)]
```

Comment:
Always put safe domain first:

```suggestion
            domain = [('opt_out', '=', True)] + domain
```